GAWB-1009 Namespace Attributes for Library Support

jmthibault79 opened the pull request:
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  - Acceptance criteria exists and is met
  - Note any changes to implementation from the description
  - Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  - If PR includes new or changed db queries, include the explain plans in the description
  - Make sure liquibase is updated if appropriate
  - If doing a migration, take a backup of the dev and alpha DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  - Authentication
  - Authorization
  - Encryption
  - Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell ![](http://i.imgur.com/9dLzbPd.png) that the PR exists if he wants to look at it **(apply requires_doge label)**
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- Review cycle:
  - LR reviews
  - Rest of team may comment on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH
  - Submitter updates documentation as needed
  - Submitter **reassigns to LR** for further feedback
- [ ] ![](http://i.imgur.com/9dLzbPd.png) sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

jmthibault79 committed:
Add unique indices to attr tables (already in liquibase)

jmthibault79 commented:
Third attempt.  Now using a string column implementation with no foreign keys.

This PR replaces the `String` representation of attribute names with the  `AttributeName` case class consisting of strings for both name and namespace, like `WorkspaceName`.

It also adds an `AttributeMap` type as an alias for the often-used `Map[String, Attribute]` construction, which is now `Map[AttributeName, Attribute]`.

Plenty of misc cleanup as well.

I have not addressed these interactions as part of this PR:
- Cromwell / WDL
- Method Configs / Method Repo
- Curator Role (GAWB-985)

jmthibault79 commented on src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponentSpec.scala:
```diff
@@ -221,11 +223,12 @@ class EntityComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers {
   }
 
-  it should "update an entity's attributes many times concurrently" in withDefaultTestDatabase {
+  // TODO: re-enable as part of GAWB-925
```
note

jmthibault79 commented on src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParsing.scala:
```diff
@@ -113,34 +116,43 @@ trait SlickExpressionParser extends JavaTokenParsers {
   private def entityRootDot: Parser[RootFunc] =
     "this." ^^ { _ => entityRootFunc}
 
+  private val namespacedIdent: Parser[String] = rep1sep(ident, ":") ^^ {
```
The `rep1sep` parser will match strings like `one:two:three` but it will then not match the cases below. I haven't tested the implications of this yet.

Could also possibly write this as a `Parser[AttributeName]` ? Again, not sure of the implications.

dvoet replied:
I would say you need a 
case _ => throw something

jmthibault79 replied:
Update: works now with `Parser[AttributeName]` and throws a `RawlsException` for `one:two:three`

jmthibault79 commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:
```diff
@@ -444,31 +452,31 @@ trait WorkflowComponent {
+      val submissionValues = resolutionsByInput map { case (inputName, recTuples: Seq[(SubmissionValidationRecord, Option[SubmissionAttributeRecord])]) =>
+        val attr = if (recTuples.forall { case (submissionRec, attrRecOpt) => attrRecOpt.isDefined }) {
+          //all attributes are real
+          val attrRecsWithRefs = recTuples map { case (rec, Some(attrOpt)) => ((workflowId, attrOpt), None) }
```
This gives a "match may not be exhaustive" warning, but we know these are all defined based on the above check.

jmthibault79 commented:
Previous versions of this PR are #526 and #530

jmthibault79 commented:
This PR does not fully address the Acceptance Criteria for GAWB-1009, in that it does not add any `isCurator` permissions checks.  Users may specify any arbitrary namespace for all attributes.

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:
```diff
@@ -136,18 +136,26 @@ trait WorkflowComponent {
+          submissionAttributeQuery.marshalAttribute(ownerId, inputResolution.inputName, attribute, Map.empty)
+        }
+
+        submissionAttributeQuery.batchInsertAttributes(attributeRecs)
```
is this just to get rid of the flatten?

jmthibault79 replied:
Yes, I may have gotten a little carried away here.

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParsing.scala:
```diff
@@ -113,34 +116,43 @@ trait SlickExpressionParser extends JavaTokenParsers {
   private def entityRootDot: Parser[RootFunc] =
     "this." ^^ { _ => entityRootFunc}
 
+  private val namespacedIdent: Parser[String] = rep1sep(ident, ":") ^^ {
```
why wouldn't this of type Parser[AttributeName]? Then parsing is not required below and pattern matching can be used.

jmthibault79 commented:
This is not ready for merge: I have found a bug when a submission writes its outputs as workspace attributes

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParsing.scala:
```diff
@@ -113,34 +108,42 @@ trait SlickExpressionParser extends JavaTokenParsers {
+  // matches some_attr_namespace:attr_name or attr_name
+  private val attributeIdent: Parser[AttributeName] = opt(ident ~ ":") ~ ident ^^ {
+    case Some(namespace ~ ":") ~ name => AttributeName(namespace, name)
+    case None ~ name => AttributeName("default", name)
```
AttributeName.defaultNamespace instead of hard coded "default"

dvoet reviewed (approved):
minor change then 👍

jmthibault79 committed:
Add Attribute Namespace to enable library

Liquibase updates:
- Add namespace to createTempTables stored procedure
- rm SUBMISSION_ATTRIBUTE_TEMP from create temp tables
- rm SUBMISSION_VALIDATION namespace

jmthibault79 commented:
I fixed the input/output resolutions bug and added tests, so I'm requesting a re-review.  I have also removed the namespace column from the Submission Validation table, where it was not appropriate.

jmthibault79 commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:
```diff
@@ -136,18 +136,32 @@ trait WorkflowComponent {
+        val attributeRecs = attributes flatMap { case (workflow, inputResolution, attribute) =>
+          val ownerId = inputResolutionRecs(workflow.workflowEntity, inputResolution.inputName).id
+
+          // TODO fix this!
```
This is out of scope for this PR, and requires some discussion.  I can write a JIRA story to refactor this if we think it's worthwhile.

dvoet replied:
they are stored as attributes because they can be different types and the attribute system works with types nicely. Another reason this makes sense is that the source of the resolutions are attributes so it makes sense to store them the same way. But I agree that namespace does not make much sense here... unless we want to do things like differentiate regular inputs from runtime, etc.

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:
```diff
@@ -136,18 +136,32 @@ trait WorkflowComponent {
+
+          // TODO fix this!
+          // this is an Input Resolution, not an Attribute Name.  Why are we storing it as an attribute?
+          val attrNameHack = AttributeName("default", inputResolution.inputName)
```
user AttributeName.defaultNamespace instead of "default"

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:
```diff
@@ -451,24 +465,27 @@ trait WorkflowComponent {
+
+          // TODO fix this!
+          // this is an Input Resolution, not an Attribute Name.  Why are we storing it as an attribute?
+          val attrNameHack = AttributeName("default", inputName)
```
AttributeName.defaultNamespace here too. I am also not convinced this is a hack.

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala:
```diff
@@ -141,7 +142,7 @@ case class SubmissionValidationHeader(
   value: Option[Attribute],
   error: Option[String],
-  inputName: String
+  inputName: String   // an expression, not an AttributeName
```
it is not an expression, it is the input name as specified in wdl

dvoet reviewed (approved):
there are many places where AttributeName.defaultNamespace should be used instead of "default". I marked those in production code but there are many more in tests.

Other than that 👍

MatthewBemis reviewed (approved):
looks good to me :+1:
